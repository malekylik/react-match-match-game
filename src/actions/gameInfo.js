import {
    CHANGE_GAME_INFO,
    INCREMENT_TIME,
    START_TIMER,
    STOP_TIMER,
    SET_GAME_ENDED,
    SET_CARDS,
} from '../constants/actions';

export const changeGameInfo = (gameDif, cardTopSrc, time) => {
    const payload = {};

    if (gameDif !== undefined) {
        payload.gameDif = gameDif;
    }

    if (cardTopSrc !== undefined) {
        payload.cardTopSrc = cardTopSrc;
    }

    if (time !== undefined) {
        payload.time = time;
    }

    return {
        type: CHANGE_GAME_INFO,
        payload
    };
};

export const incrementTime = () => ({
    type: INCREMENT_TIME,
});

export const startTimer = () => ({
    type: START_TIMER,
});

export const stopTimer = () => ({
    type: STOP_TIMER,
});

export const setCards = (cards) => ({
    type: SET_CARDS,
    payload: cards,
});

export const setGameEnded = (isEnded) => ({
    type: SET_GAME_ENDED,
    payload: isEnded,
})
