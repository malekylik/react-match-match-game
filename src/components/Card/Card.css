.card > img {
    width: 200px;
    height: 300px;
    animation-duration: .75s;
    animation-timing-function: linear;
    user-select: none;
}

.card {
    display: inline-block;
    position: relative;
    margin: 0;
    justify-self: center;
    animation-duration: 0.3s;
    animation-fill-mode: forwards;
}

.card-side-bottom {
    opacity: 0;
}

.card-side-bottom-turned {
    opacity: 1;
    transform: rotate3d(0,1,0,0deg);
    animation-direction: reverse;
    z-index: -1;
}

.card-side-top-turned {
    opacity: 0;
    transform: rotate3d(0,1,0,180deg);
    animation-direction: reverse;
    z-index: -1;
}

.card-side-bottom-turn {
    animation-name: card-turn-bottom; 
} 

.card-side-top {
    position: absolute;
    top: 0;
    left: 0;
}

.card-side-top-turn {
    animation-name: card-turn-top; 
}

.card-rotate-animation {
    animation-name: card-turn;
}

.card-disappear-animation {
    cursor: default;
    animation-name: card-disappear;
}

@keyframes card-turn-top {
    0% {
        transform: rotate3d(0,1,0,0); 
        opacity: 1;
    }

    49.99% {
        opacity: 1;
    }

    50% {
        opacity: 0;
    }

    100% {
        opacity: 0;
        transform: rotate3d(0,1,0,180deg);
    }
}

@keyframes card-turn-bottom {
    0% {
        transform: rotate3d(0,1,0,180deg); 
        opacity: 0;
    }

    49.99% {
        opacity: 0;
    }

    50% {
        opacity: 1;
    }

    100% {
        opacity: 1;
        transform: rotate3d(0,1,0,0deg);
    }
}

@keyframes card-disappear{
    to {
        opacity: 0;
    }
}
