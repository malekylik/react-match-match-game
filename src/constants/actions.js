export const CHANGE_USER_INFO = 'CHANGE_USER_INFO';

export const CHANGE_GAME_INFO = 'CHANGE_GAME_INFO';

export const START_TIMER = 'START_TIMER';
export const INCREMENT_TIME = 'INCREMENT_TIME';
export const STOP_TIMER = 'STOP_TIMER';
export const SET_CARDS = 'SET_CARDS';
export const SET_GAME_ENDED = 'SET_GAME_ENDED';

export const ADD_CARD_TO_SELECTED_CARDS = 'ADD_CARD_TO_SELECTED_CARDS';
export const RESET_SELECTED_CARDS = 'RESET_SELECTED_CARDS';
export const RESET_MATCHED_CARDS = 'RESET_MATCHED_CARDS';
export const ADD_CARDS_TO_MATCHED_CARDS = 'ADD_CARDS_TO_MATCHED_CARDS';

export const SET_TOP_STATS = 'SET_TOP_STATS';
export const SET_TOP_STATS_LOADING = 'SET_TOP_STATS_LOADING';
