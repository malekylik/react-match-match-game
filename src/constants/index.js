export const DIF = {
    LOW: {
        text: 'Low (5x2)',
        number: 5 * 2,
    },
    MEDIUM: {
        text: 'Medium (6x3)',
        number: 6 * 3,
    },
    HIGH: {
        text: 'Hight (8x3)',
        number: 8 * 3,
    },
};

export const PATH = {
    ASSETS: {
        toString() {
            return './src/imgs';
        },
        valueOf() {
            return './src/imgs';
        }
    },
};

PATH.ASSETS.CARDS = `${PATH.ASSETS}/cards`;

export const CARDS_TOP = [
    `${PATH.ASSETS.CARDS}/back1.jpg`,
    `${PATH.ASSETS.CARDS}/back2.jpg`,
    `${PATH.ASSETS.CARDS}/back3.jpg`,
];

export const CARDS_BOTTOM = [
    `${PATH.ASSETS.CARDS}/2d.jpg`,
    `${PATH.ASSETS.CARDS}/2h.jpg`,
    `${PATH.ASSETS.CARDS}/2s.jpg`,
    `${PATH.ASSETS.CARDS}/3c.jpg`,
    `${PATH.ASSETS.CARDS}/4c.jpg`,
    `${PATH.ASSETS.CARDS}/4s.jpg`,
    `${PATH.ASSETS.CARDS}/6c.jpg`,
    `${PATH.ASSETS.CARDS}/6d.jpg`,
    `${PATH.ASSETS.CARDS}/6s.jpg`,
    `${PATH.ASSETS.CARDS}/7d.jpg`,
    `${PATH.ASSETS.CARDS}/7h.jpg`,
    `${PATH.ASSETS.CARDS}/8c.jpg`,
    `${PATH.ASSETS.CARDS}/8s.jpg`,
    `${PATH.ASSETS.CARDS}/9d.jpg`,
    `${PATH.ASSETS.CARDS}/9h.jpg`,
    `${PATH.ASSETS.CARDS}/10c.jpg`,
    `${PATH.ASSETS.CARDS}/10d.jpg`,
    `${PATH.ASSETS.CARDS}/10h.jpg`,
    `${PATH.ASSETS.CARDS}/10s.jpg`,
    `${PATH.ASSETS.CARDS}/ac.jpg`,
    `${PATH.ASSETS.CARDS}/as.jpg`,
    `${PATH.ASSETS.CARDS}/jc.jpg`,
    `${PATH.ASSETS.CARDS}/jd.jpg`,
    `${PATH.ASSETS.CARDS}/jh.jpg`,
    `${PATH.ASSETS.CARDS}/js.jpg`,
    `${PATH.ASSETS.CARDS}/kd.jpg`,
    `${PATH.ASSETS.CARDS}/qh.jpg`,
];

export const RULES = [
    'Select two cards to try to match the pictures.',
    'If you match the pictures you can go again.',
    'If they don\'t match it is the computer turn them.',
    'The player that finds all pairs wins!',
    'Have Fun!',
];
