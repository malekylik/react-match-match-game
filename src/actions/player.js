import { push } from 'react-router-redux';

import { stopTimer, setGameEnded } from './gameInfo';
import {
    ADD_CARD_TO_SELECTED_CARDS,
    RESET_SELECTED_CARDS,
    RESET_MATCHED_CARDS,
    ADD_CARDS_TO_MATCHED_CARDS,
} from '../constants/actions';
import {
    getSelectedCards,
    getMatchedCards,
} from '../selectors/player';
import { getPlayerInfo } from '../selectors/playerInfo';
import { 
    getGameDif,
} from '../selectors/gameInfo';
import { loadTop, setTopStatsLoading } from './topStats';
import { saveToDatabase } from '../utils/api';

export const addCardToSelectedCards = (card) => ({
    type: ADD_CARD_TO_SELECTED_CARDS,
    payload: card,
});

export const resetSelectedCards = () => ({
    type: RESET_SELECTED_CARDS,
});

export const resetMatcherCards = () => ({
    type: RESET_MATCHED_CARDS,
});

export const addCardsToMatchedCards = (cards) => ({
    type: ADD_CARDS_TO_MATCHED_CARDS,
    payload: cards,
});

export const matchCards = () => {
    return (dispatch, getState) => {
        const state = getState();
        const [firstCard, secondCard] = getSelectedCards(state);
        const matchedCardsCount = getMatchedCards(state).length;
        const difficulty = getGameDif(state);

        const isMatch = firstCard.value === secondCard.value;

        if (isMatch) {
            firstCard.match();
            secondCard.match();
            dispatch(addCardsToMatchedCards([firstCard, secondCard]));

            if (difficulty === (matchedCardsCount + 2)) {
                dispatch(endGame());
            }
        } else {
            firstCard.turn();
            secondCard.turn();
        }

        dispatch(resetSelectedCards());
    };
};

export const endGame = () => {
    return (dispatch, getState) => {
        dispatch(stopTimer());

        const state = getState();
        const playerInfo = getPlayerInfo(state);
        const difficulty = getGameDif(state);

        dispatch(push('/result'));

        dispatch(setGameEnded(true));
        dispatch(saveResult(`${playerInfo.firstName} ${playerInfo.lastName}`, playerInfo.email, difficulty));
    }
};

export const saveResult = (name, email, score) => {
    return async (dispatch) => {
        dispatch(setTopStatsLoading(true));

        await saveToDatabase(name, email, score);
        await dispatch(loadTop());

        dispatch(setTopStatsLoading(false));
    };
};
