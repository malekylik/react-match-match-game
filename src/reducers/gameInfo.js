import { Map } from 'immutable';

import {
    CHANGE_GAME_INFO,
    INCREMENT_TIME,
    START_TIMER,
    STOP_TIMER,
    SET_CARDS,
    SET_GAME_ENDED,
} from '../constants/actions';
import { DIF, CARDS_TOP } from '../constants';

export const initialState = Map({
    gameDif: DIF.LOW.number,
    cardTopSrc: CARDS_TOP[0],
    time: 0,
    runTimer: false,
    isGameEnded: false,
    cards: [],
});

const gameInfo = (state = initialState, action) => {
    switch (action.type) {
        case CHANGE_GAME_INFO:
            return state.merge(action.payload);
        case INCREMENT_TIME:
            return state.set('time', state.get('time') + 1);
        case START_TIMER:
            return state.set('runTimer', true);
        case STOP_TIMER:
            return state.set('runTimer', false);
        case SET_CARDS:
            return state.set('cards', [...(action.payload)]);
        case SET_GAME_ENDED:
            return state.set('isGameEnded', action.payload);
        default: return state;
    }
};

export const getFormatedTime = (state) => {
    const time = state.get('time');

    let seconds = time % 60;
    let minutes = Math.floor(time / 60);
    let hours = Math.floor(time / 3600);

    if (seconds < 10) {
        seconds = `0${seconds}`;
    }

    if (minutes < 10) {
        minutes = `0${minutes}`;
    }

    if (hours < 10) {
        hours = `0${hours}`;
    }

    return {
        hours: String(hours),
        minutes: String(minutes),
        seconds: String(seconds),
    };
};

export const timerIsRun = state => state.get('runTimer');
export const getGameDif = state => state.get('gameDif');
export const getCardTopSrc = state => state.get('cardTopSrc');
export const getCards = state => state.get('cards');
export const isGameEnded = state => state.get('isGameEnded');

export default gameInfo;
