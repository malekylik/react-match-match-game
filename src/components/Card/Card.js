import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';

import CSSClassList from '../../utils/CSSClassList';

import { getSelectedCards, isCardMatched } from '../../selectors/player';
import {
    addCardToSelectedCards,
    resetSelectedCards,
    addCardsToMatchedCards,
    matchCards,
} from '../../actions/player';

import './Card.css';

class Card extends Component {
    state = {
        className: 'card',
        topSideClassName: 'card-side-top',
        bottomSideClassName: 'card-side-bottom',
    }

    isSelected = false;
    isTurning = false;
    isMatched = this.props.isMatched;

    componentWillUnmount = () => {
        clearTimeout(this.cardTurnTimeout);
        clearTimeout(this.cardDisappearTimeout);
    }

    match = () => {
        this.isMatched = true;

        this.cardDisappearTimeout = setTimeout(() => {
            this.setState({ className: new CSSClassList(this.state.className).add('card-disappear-animation') });
        }, 1250);
    }

    turn = () => {
        this.cardTurnTimeout = setTimeout(() => {
            this.clickHander();
            this.isSelected = false;
        }, 1250);
    }

    select = () => {
        if ((!this.isSelected && !this.isMatched) && this.props.selectedCards.length < 2
            && !this.isTurning) {
            this.clickHander();

            this.isSelected = true;

            this.props.addCardToSelectedCards({
                value: this.props.value,
                match: this.match,
                turn: this.turn,
            });

            if (this.props.selectedCards.length > 0) {
                this.props.matchCards();
            }
        }
    }

    clickHander = () => {
        if (!this.isTurning) {
            this.setState({
                topSideClassName: new CSSClassList(this.state.topSideClassName).add('card-side-top-turn'),
                bottomSideClassName: new CSSClassList(this.state.bottomSideClassName).add('card-side-bottom-turn'),
            });

            this.isTurning = true;
        }
    }

    animationEndHandler = () => {
        if (this.isTurning) {
            let topSideClassList = new CSSClassList(this.state.topSideClassName);
            let bottomSideClassName = new CSSClassList(this.state.bottomSideClassName);

            topSideClassList.toggle('card-side-top-turned');
            bottomSideClassName.toggle('card-side-bottom-turned');

            topSideClassList.remove('card-side-top-turn');
            bottomSideClassName.remove('card-side-bottom-turn');

            this.setState({
                topSideClassName: String(topSideClassList),
                bottomSideClassName: String(bottomSideClassName),
            });

            this.isTurning = false;
        }
    }

    render() {
        const { bottomSideSrc, topSideSrc } = this.props;
        const isShouldCardBeVisible = !(!this.isTurning && !this.cardDisappearTimeout && this.props.isMatched);
        const style = {
            opacity: isShouldCardBeVisible ? 1 : 0,
            cursor: !this.props.isMatched ? 'pointer' : 'default',
        };

        return (
            <figure style={style} className={this.state.className} onClick={this.select} onAnimationEnd={this.animationEndHandler} >
                <img style={{ cursor: this.state.cursor }} className={this.state.bottomSideClassName} src={bottomSideSrc} />
                <img style={{ cursor: this.state.cursor }} className={this.state.topSideClassName} src={topSideSrc} />
            </figure>
        )
    }
}

Card.propTypes = {
    bottomSideSrc: PropTypes.string.isRequired,
    topSideSrc: PropTypes.string.isRequired,
    value: PropTypes.number.isRequired,
};

export default connect(
    (state, ownProps) => ({
        selectedCards: getSelectedCards(state),
        isMatched: isCardMatched(state, ownProps.value),
    }), {
        addCardToSelectedCards,
        resetSelectedCards,
        addCardsToMatchedCards,
        matchCards,
    })(Card);
