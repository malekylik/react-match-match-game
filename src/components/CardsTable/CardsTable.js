import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';

import { startTimer, stopTimer } from '../../actions/gameInfo';
import { resetSelectedCards } from '../../actions/player';
import { getCards, isGameEnded } from '../../selectors/gameInfo';
import { getMatchedCards } from '../../selectors/player';

import './CardsTable.css';

class CardsTable extends Component {
    componentWillUnmount() {
        this.props.stopTimer();
        this.props.resetSelectedCards();
    }

    componentDidMount() {
        if (!this.props.isGameEnded) {
            this.props.startTimer();
        }
    }

    render() {
        const { cards } = this.props;

        return (
            <section className="cards-table">
                {cards}
            </section>
        );
    }
}

CardsTable.propTypes = {
    cards: PropTypes.oneOfType([
        PropTypes.element,
        PropTypes.arrayOf(PropTypes.element)
    ]),
};

export default connect((state) => ({
    cards: getCards(state),
    matchedCards: getMatchedCards(state),
    isGameEnded: isGameEnded(state),
}), {
        startTimer,
        stopTimer,
        resetSelectedCards,
    })(CardsTable);
