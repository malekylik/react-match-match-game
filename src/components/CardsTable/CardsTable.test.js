import React from 'react';
import configureStore from 'redux-mock-store';

import { Provider } from 'react-redux';
import { render, mount } from 'enzyme';

import CardsTable from './CardsTable';
import Card from '../Card/Card';

import { START_TIMER } from '../../constants/actions';

import { initialState as playerInitialState } from '../../reducers/player';
import { initialState as gameInfoInitialState } from '../../reducers/gameInfo';
import { CARDS_TOP, CARDS_BOTTOM } from '../../constants';
import { stopTimer } from '../../actions/gameInfo';
import { resetSelectedCards } from '../../actions/player';

const middlewares = [];
const mockStore = configureStore(middlewares);

describe('CardsTable', () => {
    const initialState = {
        player: playerInitialState,
        gameInfo: gameInfoInitialState.set('cards', [<Card key={0} value={0} bottomSideSrc={CARDS_BOTTOM[0]} topSideSrc={CARDS_TOP[0]} />]),
    };

    let store;

    beforeEach(() => {
        store = mockStore(initialState);
    });

    it('Should match snapshot with data', () => {
        const wrapper = render(
            <Provider store={store}>
                <CardsTable />
            </Provider>
        );

        expect(wrapper).toMatchSnapshot();
    });

    describe('Start timer', () => {
        it('Should start time when it is mounted', () => {
            mount(
                <Provider store={store}>
                    <CardsTable />
                </Provider>
            );

            const actions = store.getActions();
            const expectedPayload = { type: START_TIMER }

            expect(actions).toEqual([expectedPayload]);
        });

        it('Shouldn\'t start time if game is ended', () => {
            initialState.gameInfo = initialState.gameInfo.set('isGameEnded', true);
            store = mockStore(initialState);

            mount(
                <Provider store={store}>
                    <CardsTable />
                </Provider>
            );

            const actions = store.getActions();
            expect(actions).toEqual([]);
        });
    });

    it('Should stop time and reset selected cards when it is unmounted', () => {
        const wrapper = mount(
            <Provider store={store}>
                <CardsTable />
            </Provider>
        );

        wrapper.unmount();

        const actions = store.getActions();
        const stopTimerAction = stopTimer();
        const resetSelectedCardsAction = resetSelectedCards();

        expect(~(actions.findIndex(action => action.type === stopTimerAction.type))).toBeTruthy();
        expect(~(actions.findIndex(action => action.type === resetSelectedCardsAction.type))).toBeTruthy();
    });
});
