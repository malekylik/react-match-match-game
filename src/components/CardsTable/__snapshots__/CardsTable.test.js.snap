// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardsTable Should match snapshot with data 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "cards-table",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "card",
          "style": "opacity:1;cursor:pointer",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "card-side-bottom",
              "src": "./src/imgs/cards/2d.jpg",
            },
            "children": Array [],
            "name": "img",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": "card-side-top",
                "src": "./src/imgs/cards/back1.jpg",
              },
              "children": Array [],
              "name": "img",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "src": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "src": undefined,
              },
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
              "src": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
              "src": undefined,
            },
          },
          Object {
            "attribs": Object {
              "class": "card-side-top",
              "src": "./src/imgs/cards/back1.jpg",
            },
            "children": Array [],
            "name": "img",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "class": "card-side-bottom",
                "src": "./src/imgs/cards/2d.jpg",
              },
              "children": Array [],
              "name": "img",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "src": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "src": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
              "src": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
              "src": undefined,
            },
          },
        ],
        "name": "figure",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "style": undefined,
        },
      },
    ],
    "name": "section",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
